clog.3
.\"
.\" Copyright (c) 2010 Marco Peereboom <marco@peereboom.us>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: October 10 2011 $
.Dt CLOG 3
.Os
.Sh NAME
.Nm clog
.Nd Copious Logger
.Sh SYNOPSIS
.Fd #include <clog.h>
.Ft void
.Fn clog_init "int no_syslog"
.Ft int
.Fn clog_set_flags "uint32_t flags"
.Ft uint32_t
.Fn clog_get_flags "void"
.Ft int
.Fn clog_set_logfile "const char *logfile"
.Ft const char *
.Fn clog_get_logfile "void"
.Ft void
.Fn clog_set_mask "uint64_t mask"
.Ft uint64_t
.Fn clog_get_mask "void"
.Ft int
.Fn clog_mask_is_set "uint64_t bit"
.Ft void
.Fn clog_version "int *major" "int *minor" "int *patch"
.Ft void
.Fn CABORT "const char *format" "..."
.Ft void
.Fn CABORTX "const char *format" "..."
.Ft void
.Fn CALERT "const char *format" "..."
.Ft void
.Fn CALERTX "const char *format" "..."
.Ft void
.Fn CCRIT "const char *format" "..."
.Ft void
.Fn CCRITX "const char *format" "..."
.Ft void
.Fn CEMERG "const char *format" "..."
.Ft void
.Fn CEMERGX "const char *format" "..."
.Ft void
.Fn CERR "const char *format" "..."
.Ft void
.Fn CERRX "const char *format" "..."
.Ft void
.Fn CFATAL "const char *format" "..."
.Ft void
.Fn CFATALX "const char *format" "..."
.Ft void
.Fn CINFO "const char *format" "..."
.Ft void
.Fn CNINFO "uint64_t mask"  "const char *format" "..."
.Ft void
.Fn CDBG "const char *format" "..."
.Ft void
.Fn CNDBG "uint64_t mask" "const char *format" "..."
.Ft void
.Fn CNOTICE "const char *format" "..."
.Ft void
.Fn CNNOTICE "uint64_t mask" "const char *format" "..."
.Ft void
.Fn CWARN "const char *format" "..."
.Ft void
.Fn CNWARN "uint64_t mask" "const char *format" "..."
.Ft void
.Fn CWARNX "const char *format" "..."
.Ft void
.Fn CNWARNX "uint64_t mask" "const char *format" "..."
.Sh DESCRIPTION
.Fn clog_init
prepares the library for first use.
This function should be called before any other function is called.
.Pp
.Fn clog_set_flags
sets the operational flags for
.Nm .
Possible values are:
.Bl -tag -width "CLOG_F_ENABLE" -offset indent -compact
.It Cm CLOG_F_FILE
Log source file name.
.It Cm CLOG_F_FUNC
Log source function name.
.It Cm CLOG_F_LINE
Log source file line number.
.It Cm CLOG_F_PID
Log program name and PID.
.It Cm CLOG_F_DATE
Log date in
.Xr ctime 3
format.
.It Cm CLOG_F_DTIME
Log delta time from previous log function invocation.
This time delta has micro second resolution.
.It Cm CLOG_F_FILOG
Log to
.Fa logfile
which is given as an argument to
.Fn clog_set_logfile .
The
.Fn clog_set_logfile
function must be called before setting this flag.
.It Cm CLOG_F_STDERR
Log to
.Fa stderr .
.It Cm CLOG_F_SYSLOG
Log to
.Xr syslog 3 .
Note that
.Xr syslog 3
always prepends the
.Xr ctime 3
date.
Currently all
.Nm
output goes to
.Fa LOG_DAEMON .
.It Cm CLOG_F_ENABLE
Enable logging.
.It Cm CLOG_F_DBGENABLE
Enable debug logging.
.El
.Pp
Only supported flags can be set and this function will return 0 on success and
1 on failure.
The currently set flags may be fetched with
.Fn clog_get_flags .
.Pp
.Fn clog_set_logfile
sets the log file to log to when
.Fa CLOG_F_FILOG
is set.  This function must be called before setting
.Fa CLOG_F_FILOG .
The returned value is 0 on success and 1 on failure.
.Pp
.Fn clog_set_mask
sets external debug mask.
The debug mask must be a bitfield and must be defined by the caller.
It is used to control what to log and what to skip.
The current debug mask may be fetched with
.Fn clog_get_mask ,
alternatively whether a bit is set may be tested with
.Fn clog_mask_is_set .
.Pp
.Fn clog_version
gets the current version of the library at run time.
.Pp
All logging functions use the format as defined by
.Xr printf 3 .
The
.Fa N
in the name of the logging functions designates the use of a mask.
The functions for logging important messages (
.Fn CABORT
.Fn CALERT ,
.Fn CCRIT ,
.Fn CEMERG ,
.Fn CERR
and
.Fn CFATAL
) do not have masked variants since they are intended only for important
messages which should require no masking.
.Pp
.Fn CALERT ,
.Fn CCRIT ,
.Fn CDBG ,
.Fn CEMERG
.Fn CERR ,
.Fn CINFO ,
.Fn CNOTICE ,
.Fn CWARN ,
(and their masked counterparts where present) will log informational
messages with severities corresponding to the well known
.Xr syslog 3
levels if the
.Fa CLOG_F_ENABLE
flag is present and they are unmasked.
The
.Fn CDBG
family of functions additionally require the
.Fa CLOG_F_DBGENABLE
flag to be set.
.Pp
.Fn CFATAL
and
.Fn CFATALX
terminate the program using
.Xr exit 3
with a non-zero return value.
.Pp
.Fn CABORT
and
.Fn CABORTX
terminate the program using
.Xr abort 3
.Pp
The functions with an
.Fa X
in their name suppress
.Xr strerror 3
output, and are functionally equivalent to the
.Xr err 3
and
.Xr warn 3
family of functions.
.Sh SEE ALSO
.Xr printf 3
.Xr ctime 3
.Xr syslog 3
.Xr err 3
.Xr warn 3
.Sh HISTORY
.An -nosplit
.Nm
was written by
.An Marco Peereboom Aq marco@peereboom.us


clog.c
/* $OpenBSD: log.c,v 1.8 2007/08/22 21:04:30 ckuethe Exp $ */
/*
 * Copyright (c) 2010 Marco Peereboom <marco@peereboom.us>
 * Copyright (c) 2003, 2004 Henning Brauer <henning@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF MIND, USE, DATA OR PROFITS, WHETHER
 * IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "clog.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <stdarg.h>
#include <inttypes.h>
#include <limits.h>
#include <time.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>

#ifdef BUILDSTR
static const char *vertag = "version: " CLOG_VERSION " " BUILDSTR;
#else
static const char *vertag = "version: " CLOG_VERSION;
#endif

extern char		*__progname;

static int		clog_initialized;
static uint64_t		clog_ext_mask;
static uint32_t		clog_flags;
static struct timeval	clog_start_of_day;
static char		clog_logfile[PATH_MAX] = "";

int
clog_set_logfile(const char *logfile)
{
	if (strlcpy(clog_logfile, logfile, sizeof(clog_logfile)) >=
		sizeof(clog_logfile))
		return (1);
	return (0);
}

const char *
clog_get_logfile(void)
{
	return (clog_logfile);
}

const char *
clog_verstring(void)
{
	return (vertag);
}

void
clog_version(int *major, int *minor, int *patch)
{
	*major = CLOG_VERSION_MAJOR;
	*minor = CLOG_VERSION_MINOR;
	*patch = CLOG_VERSION_PATCH;
}

void
clog_opensyslog(void)
{
	openlog(__progname, LOG_PID | LOG_NDELAY | LOG_CONS, LOG_DAEMON);
}

void
clog_initialize(int log_syslog)
{
	clog_ext_mask = 1llu<<63;
	gettimeofday(&clog_start_of_day, NULL);

	if (log_syslog)
		clog_opensyslog();

	tzset();

	clog_initialized = 1;
}

void
clog_end(void)
{
	if (clog_flags & CLOG_F_SYSLOG)
		closelog();
	clog_initialized = 0;
}

int
clog_set_flags(uint32_t f)
{
	if ((f & ~CLOG_F_ALLFLAGS) != 0)
		return (1);

	if ((f & CLOG_F_FILOG) != 0 && clog_logfile[0] == '\0')
		return (1);

	if ((f & CLOG_F_SYSLOG) != (clog_flags & CLOG_F_SYSLOG)) {
		/* syslog toggle */
		if ((clog_flags & CLOG_F_SYSLOG))
			closelog();
		else
			clog_opensyslog();
	}

	clog_flags = f;

	return (0);
}

uint32_t
clog_get_flags(void)
{
	return (clog_flags);
}

void
clog_set_mask(uint64_t f)
{
	clog_ext_mask = f | (1llu<<63);
}

uint64_t
clog_get_mask(void)
{
	return (clog_ext_mask & ~(1llu<<63));
}

int
clog_mask_is_set(uint64_t bit)
{
	return (clog_ext_mask & bit);
}

void
clog_print(int pri, int do_errno, const char *file, const char * func, int line,
    const char *fmt, va_list ap)
{
	const char		*fi = "", *fu = "", *li = "";
	const char		*pi2 = "", *fi2 = "", *fu2 = "", *li2 = "";
	char			li_buf[16], *start = "", *end = "";
	char			*s = NULL, *ts = "", *ts2 = "" , delta[32];
	char			*sl = NULL, *er = "", *er2 = "", *pi = "";
	int			got_some = 0, free_pi = 0;
	struct timeval		now, elapsed;
	time_t			tnow;
	va_list			fap, sap;
	FILE			*stream;

	delta[0] = '\0';
	if ((CLOG_F_DTIME & clog_flags) != 0) {
		if (clog_initialized == 0)
			clog_initialize(0);

		if (gettimeofday(&now, NULL) != -1) {
			timersub(&now, &clog_start_of_day, &elapsed);
			snprintf(delta, sizeof delta, "%" PRId64 ".%.6ld: ",
			    (int64_t) elapsed.tv_sec, (long) elapsed.tv_usec);
		}
	}

	if ((CLOG_F_DATE & clog_flags) != 0) {
		time(&tnow);
		ts = ctime(&tnow) + 4;
		ts[15] = '\0';
		if ((CLOG_F_DTIME & clog_flags) == 0)
			ts2 = ": ";
		else
			ts2 = " ";
	}

	if ((CLOG_F_PID & clog_flags) != 0) {
		if (asprintf(&pi, "%s(%d)", __progname, getpid()) != -1) {
			pi2 = " ";
			got_some = 1;
			free_pi = 1;
		} else {
			pi = "";
		}
	}

	if ((CLOG_F_FILE & clog_flags) != 0) {
		fi = file;
		fi2 = " ";
		got_some = 1;
	}
	if ((CLOG_F_FUNC & clog_flags) != 0) {
		fu = func;
		fu2 = " ";
		got_some = 1;
	}
	if ((CLOG_F_LINE & clog_flags) != 0) {
		snprintf(li_buf, sizeof li_buf, "%d", line);
		li = li_buf;
		li2 = " ";
		got_some = 1;
	}

	if (got_some) {
		start = "< ";
		end = "> ";
	}

	if (do_errno) {
		er = ": ";
		er2 = strerror(errno);
	}

	if (asprintf(&s, "%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s\n",
	    ts, ts2, delta, start, pi, pi2, fi, fi2, fu, fu2,
	    li, li2, end, fmt, er, er2) == -1) {
		/* out of memory */
		abort();
	}

	va_copy(sap, ap);
	va_copy(fap, ap);

	if ((CLOG_F_FILOG & clog_flags) != 0) {
		stream = fopen(clog_logfile, "a");
		if (stream) {
			vfprintf(stream, s, fap);
			fclose(stream);
		}
	}

	if ((CLOG_F_STDERR & clog_flags) != 0) {
		vfprintf(stderr, s, ap);
		fflush(stderr);
	}

	if ((CLOG_F_SYSLOG & clog_flags) != 0) {
		if ((CLOG_F_DATE & clog_flags) != 0)
			sl = s + 16;
		else
			sl = s;

		if (clog_initialized == 0)
			clog_initialize(1);

		vsyslog(pri, sl, sap);
	}

	free(s);
	if (free_pi)
		free(pi);
}

void
clog_dbg_internal(int pri, int do_errno, uint64_t mask, const char *file,
    const char * func, int line, const char *fmt, ...)
{
	va_list			ap;

	if ((mask & clog_ext_mask) == 0 || (clog_flags & CLOG_F_ENABLE) == 0)
		return;
	if (pri == LOG_DEBUG && !(clog_flags & CLOG_F_DBGENABLE))
		return;

	va_start(ap, fmt);
	clog_print(pri, do_errno, file, func, line, fmt, ap);
	va_end(ap);
}

/* old interface */
int		logdebug;
int		debugsyslog;

void
vlog(int pri, const char *fmt, va_list ap)
{
	char	*nfmt;

	if (logdebug) {
		/* best effort in out of mem situations */
		if (asprintf(&nfmt, "%s\n", fmt) == -1) {
			vfprintf(stderr, fmt, ap);
			fprintf(stderr, "\n");
		} else {
			vfprintf(stderr, nfmt, ap);
			free(nfmt);
		}
		fflush(stderr);
	} else
		vsyslog(pri, fmt, ap);
}

void
logit(int pri, const char *fmt, ...)
{
	va_list	ap;

	va_start(ap, fmt);
	vlog(pri, fmt, ap);
	va_end(ap);
}

void
clog_init(int n_debug)
{
	logdebug = n_debug;

	if (!logdebug)
		openlog(__progname, LOG_PID | LOG_NDELAY | LOG_CONS,
		    LOG_DAEMON);
	tzset();

	clog_initialize(!logdebug); /* initialize new interface too */
}

void
clog_warn(const char *emsg, ...)
{
	char	*nfmt;
	va_list	 ap;

	/* best effort to even work in out of memory situations */
	if (emsg == NULL)
		logit(LOG_CRIT, "%s", strerror(errno));
	else {
		va_start(ap, emsg);

		if (asprintf(&nfmt, "%s: %s", emsg, strerror(errno)) == -1) {
			/* we tried it... */
			vlog(LOG_CRIT, emsg, ap);
			logit(LOG_CRIT, "%s", strerror(errno));
		} else {
			vlog(LOG_CRIT, nfmt, ap);
			free(nfmt);
		}
		va_end(ap);
	}
}

void
clog_warnx(const char *emsg, ...)
{
	va_list	 ap;

	va_start(ap, emsg);
	vlog(LOG_CRIT, emsg, ap);
	va_end(ap);
}

__dead void
clog_fatal(const char *emsg)
{
	if (emsg == NULL)
		logit(LOG_CRIT, "fatal: %s", strerror(errno));
	else
		if (errno)
			logit(LOG_CRIT, "fatal: %s: %s",
			    emsg, strerror(errno));
		else
			logit(LOG_CRIT, "fatal: %s", emsg);

	exit(1);
}

__dead void
clog_fatalx(const char *emsg, ...)
{
	va_list  ap;

	va_start(ap, emsg);
	vlog(LOG_CRIT, emsg, ap);
	va_end(ap);

	exit(1);
}

void
clog_info(const char *emsg, ...)
{
	va_list	 ap;

	va_start(ap, emsg);
	vlog(LOG_INFO, emsg, ap);
	va_end(ap);
}

void
clog_debug(const char *emsg, ...)
{
	va_list	 ap;

	if (logdebug || debugsyslog) {
		va_start(ap, emsg);
		vlog(LOG_DEBUG, emsg, ap);
		va_end(ap);
	}
}

const char *
clog_sockaddr(struct sockaddr *sa)
{
	static char	buf[NI_MAXHOST];

	if (getnameinfo(sa, SA_LEN(sa), buf, sizeof(buf), NULL, 0,
	    NI_NUMERICHOST))
		return ("(unknown)");
	else
		return (buf);
}
